File: cdiddy/vmbase.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "vmbase.h"

DVM setup_vmbase() {

    DVM dvm           = (DVM) malloc(sizeof(DVM_Data));

    dvm->ram          = (int*) malloc(sizeof(int*) * 4096);

    int i;
    for (i=0; i < 4096; i++) {
        dvm->ram[i] = 0;
    }

    dvm->position     = 1;
    dvm->running      = 1;
    dvm->status       = 0;
    return dvm;
}

void execute_next_instruction(DVM dvm) {
    int i = getMem(dvm, -1);
    (*dvm->instructions[i]) (dvm);
}

void set_instruction_pointer(DVM dvm, int address) {
    dvm->position = address;
}

void next(DVM dvm) {
    dvm->position++;
}

int getMem(DVM dvm, int addr) {

    int output;

    if (addr == -1) {
        output = dvm->ram[dvm->position];
    } else {
        output = dvm->ram[addr];
    }
    return output;
}

void setMem(DVM dvm, int addr, int value) {
    dvm->ram[addr] = value;
}

void system_exit(DVM dvm, int value) {
    dvm->running = 0;
    dvm->status = value;
}

void system_out(DVM dvm, int c) {
    putchar(c);
}

void load_program(DVM dvm, char *program_data) {
}

void cleanup_dvm(DVM dvm) {
    free(dvm->ram);
    free(dvm);
}

/*
int main(int argc, char *argv[]) {
    DVM dvm = (DVM) setup_vmbase();
    next(dvm);
    next(dvm);
    cleanup_dvm(dvm);

    return 0;
}
*/

File: pdiddy/dvm/vmbase.py
#!/usr/bin/env python

# This is the class that contains all the basic methods for the Diddy VM
#
# This machine operates entirely in RAM
# There are registers for special uses,
#   Input
#   Output
#   Exiting
# Instructions can have a variable number of arguments
# The machines memory width is 16 bits

import sys


class VMBase(object):

    """
    The base VM class
    Contains the internal data structures as well as general purpose methods

    A VM Class with the instruction methods should inherit this
    """

    def __init__(self, debug=False, stack_display=False):
        """
        Setup the ram, instruction dictionary and register dictionary
        """
        self.ram = dict.fromkeys(xrange(2 ** 12), 0)
        self.instructions = {}

        self.position = 1

        self.running = True
        self.status = 0

        self.stack = []

        self.debug = debug
        self.stack_display = stack_display

        self.instruction_mask = 31 << 28

    # General purpose functions
    def executeNextInstruction(self):
        """
        Retrieves the value from the memory location pointed to
        by the position pointer and executes it
        """
        bits = self.ram[self.position]
        inst = (bits & self.instruction_mask) >> 28

        self.instructions[inst](bits)

    def setInstructionPointer(self, address):
        """
        Sets the instruction pointer to the specified address
        """
        self.position = address

    def next(self):
        """
        Increments program position counter
        """
        self.position += 1

    def pushStack(self, data):
        """
        Push a value onto the VM stack
        data should be an integer
        """
        value = self.stack.append(data)
        if self.debug:
            print("stack size:  %i" % len(self.stack))
        if self.stack_display:
            print(self.stack)
        return value

    def popStack(self):
        """
        Pop a value off the VM stack
        data should be an integer
        """
        try:
            value = self.stack.pop()
            if self.debug:
                print("stack size:  %i" % len(self.stack))
            if self.stack_display:
                print(self.stack)
            return value
        except IndexError:
            self.systemExit(1)

    def getMem(self, addr=-1):
        """
        Retrieve a value from a memory address
        If no address is given then the value from the
        position pointer will be retrieved
        """
        if addr == -1:
            output = self.ram[self.position]
        else:
            output = self.ram[addr]
        return output

    def setMem(self, addr, value):
        """
        Set a value in memory
        If the memory position corresponds to a register then
        the register function will be called instead
        """
        self.ram[addr] = value

    def systemExit(self, value):
        """
        Set the running flag to false so the external program loop stops
        Also sets the exit status of the VM
        """
        self.running = False
        self.status = value

    def systemOut(self, c):
        """
        Print a char to stdout
        """
        sys.stdout.write(chr(c))

    def loadProgram(self, programData):
        """
        Loads a program into memory
        The programdata should be a list of integers
        """
        for value in programData:
            self.setMem(self.position, value)
            self.next()
        self.setInstructionPointer(1)

if __name__ == '__main__':
    import unittest
    from tests.testvmbase import TestVMBase
    unittest.main()
